python/cuml/neighbors/kernel_density.py

```diff
@@ -0,0 +1,406 @@
+
+    """
+
+    def __init__(
+        self,
+        *,
+        bandwidth=1.0,
+        kernel="gaussian",
+        metric="euclidean",
+        metric_params=None,
+        output_type=None,
+        handle=None,
+        verbose=False
+    ):
+        super(KernelDensity, self).__init__(
+            verbose=verbose, handle=handle, output_type=output_type
+        )
+        self.bandwidth = bandwidth
+        self.kernel = kernel
+        self.metric = metric
+        self.metric_params = metric_params
+
+        if bandwidth <= 0:
+            raise ValueError("bandwidth must be positive")
+        if kernel not in VALID_KERNELS:
+            raise ValueError("invalid kernel: '{0}'".format(kernel))
+
+    def get_param_names(self):
+        return super().get_param_names() + [
+            "bandwidth",
+            "kernel",
+            "metric",
+            "metric_params",
+        ]
+
+    def fit(self, X, y=None, sample_weight=None):
+        """Fit the Kernel Density model on the data.
+
+        Parameters
+        ----------
+
+        X : array-like of shape (n_samples, n_features)
+            List of n_features-dimensional data points.  Each row
+            corresponds to a single data point.
+        y : None
+            Ignored.
+        sample_weight : array-like of shape (n_samples,), default=None
+            List of sample weights attached to the data X.
+
+        Returns
+        -------
+
+        self : object
+            Returns the instance itself.
+        """
+        if sample_weight is not None:
+            self.sample_weight_ = input_to_cupy_array(
+                sample_weight, check_dtype=[cp.float32, cp.float64]
+            ).array
+            if self.sample_weight_.min() <= 0:
+                raise ValueError("sample_weight must have positive values")
+        else:
+            self.sample_weight_ = None
+
+        self.X_ = input_to_cupy_array(
+            X, order="C", check_dtype=[cp.float32, cp.float64]
+        ).array
+
+        return self
+
+    def score_samples(self, X):
+        """Compute the log-likelihood of each sample under the model.
+
+        Parameters
+        ----------
+
+        X : array-like of shape (n_samples, n_features)
+            An array of points to query.  Last dimension should match dimension
+            of training data (n_features).
+
+        Returns
+        -------
+
+        density : ndarray of shape (n_samples,)
+            Log-likelihood of each sample in `X`. These are normalized to be
+            probability densities, so values will be low for high-dimensional
+            data.
+        """
+        if not hasattr(self, "X_"):
+            raise NotFittedError()
+
+        if self.metric_params:
+            if len(self.metric_params) != 1:
+                raise ValueError(
+                    "Cuml only supports metrics with a single arg.")
+            metric_arg = list(self.metric_params.values())[0]
```

Comment:
Can you open an issue for this and reference it there? This is a bug.

Reply:
#4605